Assignment4/PointGrid.py

```diff
@@ -0,0 +1,170 @@
+class Grid:
+
+    def __init__(self, grid):
+        """
+        For grid initialization with a grid with elements.
+        :param grid: list(list(any))
+        :return: None
+        """
+        self._grid = grid
+        self._points = {}
+        self._update_points()
+        self._islands = []
+
+    class Point:
+        def __init__(self, value, coordinates):
+            self.coordinates = coordinates
+            self.value = value
+            self.is_visited = False
+
+        def visit(self):
+            """
+            Set the visit attribute to be True.
+            :return: None
+            """
+            self.is_visited = True
+
+        def __eq__(self, other):
+            """
+            Supports equivalence for the Point. Returns True if the elements self and other are equal, False otherwise.
+            :param other: any
+            :return: bool
+            """
+            if isinstance(other, self.__class__):
+                return self.coordinates == other.coordinates and self.value == other.value and self.is_visited == other.is_visited
+            else:
+                return False
+
+    def find_point(self, point_value):
+        """
+        Return the first NOT VISITED Point in self._points with the given value if exists. None otherwise.
+        :param point_value: any
+        :return: Point or None
+        """
+        for coord in self._points:
+            point = self.get_point(coord)
+            if point.value == point_value and point.is_visited == False:
+                return point
+
+    def get_point(self, coordinates):
+        """
+        Get element in Grid by its coordinates. Raises ValueError if coordinates are invalid.
+        :param coordinates: tuple(int,int)
+        :return: Point or raise ValueError
+        """
+        if not self.valid_coordinates(coordinates):
+            raise ValueError("invalid coordinates")
+        return self._points[coordinates]
+
+    def get_element(self, coordinates):
+        """
+        Return element in the Grid by its coordinates.
+        :param coordinates: tuple(int, int)
+        :return: any
+        """
+        return self._grid[coordinates[0]][coordinates[1]]
+
+    def valid_coordinates(self, coordinates):
+        """
+        Check if the coordinates are valid. Return True if so, False otherwise.
+        :param coordinates: tuple(int, int)
+        :return: bool
+        """
+        Rs, Cs = self.size()
+        return coordinates[0] in range(0, Rs) and coordinates[1] in range(0, Cs)

```

Comment:
Do you have any insight how this code might run? What is the runtime complexity of checking for valid coordinates here?

I'd suspect that `range(0, Rs)` would generate numbers from 0 and the `in` check for a generator would compare it to each element in sequence, so O(N) to the length of dimensions?

```
>>> timeit.timeit('90 in range(0, 100)')
1.1249170303344727
>>> timeit.timeit('90 >= 0 and 90 < 100')
0.04843902587890625
```

So for a 100x100 grid each `valid_coordinates` check would take order of 1s?

Premature optimization is counter-productive but if there are other equally available and readable options then they should be used.